Quote column names to avoid name colisions with reserved names or keywords

andygrunwald opened the pull request:
**What type of PR is this?**

bug

**What this PR does / why we need it**:

Imagine we have this CSV file (called `nutrient.csv`, from the https://fdc.nal.usda.gov/download-datasets.html // Supporting data for Downloads - April 2021 (CSV – 1.0M):

```
id,name,unit_name,nutrient_nbr,rank
1001,Solids,G,201,200
1002,Nitrogen,G,202,500
1003,Protein,G,203,600
1004,Total lipid (fat),G,204,800
1005,"Carbohydrate, by difference",G,205,1110
```

The command `./csv2db generate --file nutrient.csv --table nutrient --verbose` generates:

```
Finding file(s).
Found 1 file(s).
Generating CREATE TABLE statement.
CREATE TABLE nutrient
(
  ID VARCHAR(1000),
  NAME VARCHAR(1000),
  UNIT_NAME VARCHAR(1000),
  NUTRIENT_NBR VARCHAR(1000),
  RANK VARCHAR(1000)
);
```

Executing this query against a MySQL Community Server - GPL 8.0.21. will result in

```
[ERROR in query 2] You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'RANK VARCHAR(1000)
)' at line 7
```

The reason: `RANK` is not quoted and [a reserved keyword since MySQL 8.0.2](https://dev.mysql.com/doc/refman/8.0/en/keywords.html).

A similar error will happen for the insert statements.

Solution: Quoting column names.

This PR quotes the column names in the CREATE and INSERT statements.

**Special notes for your reviewer**:

This was tested against MySQL Community Server - GPL 8.0.21.
Other database types were not tested.

andygrunwald committed:
Quote column names to avoid name colisions with reserved names or keywords

gvenzl commented:
Thanks a lot @andygrunwald for this PR!

Unfortunately, I'm pretty sure that different databases use different quoting techniques.
Will have to go through the documentation of each supported database and find the correct characters to use.

andygrunwald commented:
@gvenzl Would it be a (pragmatic) solution to apply these changes only for MySQL?

Like
```python
if cfg.db_type == f.DBType.MYSQL.value:
    # Apply "`"
else:
    # normal behaviour
```

Let me know what you think

gvenzl commented:
Hey @andygrunwald,

You know, I was thinking along similar lines this morning.
Not all databases may have the same restrictions with keywords, some might be smart enough to understand when it's a column name and hence not need the quoting.

On the other hand, I know that Oracle and I believe SQL Server and Db2 will also introduce case sensitivity the moment an identifier is quoted, i.e. `SELECT "RANK" FROM nutrient` is a different column than `SELECT "rank" FROM nutrient` while `SELECT RANK FROM nutrient` and `SELECT rank FROM nutrient` is the same.

Right now `csv2db` makes all identifier uppercase as the case does not matter.
But the moment quotes come into the picture...

I also need to read up on MySQL and on what other semantics the quoting character may imply as well.
Perhaps it's not a biggy to just quote all the identifiers for all the supported databases, perhaps it's just necessary for MySQL and perhaps it's best left to another command-line option.

andygrunwald commented:
@gvenzl Sounds good and reasonable. Thanks for putting thoughts into this.

Is there anything that comes into your mind where I can help/support?

gvenzl commented:
Hey @andygrunwald,

Sorry, Monday was a bank holiday over here and I took Friday off as well to go to the middle of nowhere and recharge a bit.
Hence why I didn't reply until now.

You could help with the research for quoted identifiers if you like.
That is my next step, see what's implied with the quotes for each supported database, as described above.
We have to check the following:

1) Oracle
1) MySQL
1) Postgres
1) SQL Server
1) Db2

andygrunwald commented:
@gvenzl No worries. Take your time. Life is more important. I did not expect a super-fast response, because I know how it can be to maintain a project.

Ok, let's start with the research:

### 1. Oracle

Oracle seems to be using the double quote `"`.
I was not able to find an official documentation page about this, but a few other sources:

[Ispirer SQLWays Database Migration Software > Oracle Reserved Words](https://doc.ispirer.com/sqlways/Output/SQLWays-1-134.html):

> There are reserved words in Oracle which cannot be used as identifiers (table or column names etc.) without being delimited with double quotation marks ("). The only exception is that you cannot use the uppercase reserved word ROWID as an identifier, even in double quotation marks.

[StackOverflow: How do I escape a reserved word in Oracle?](https://stackoverflow.com/questions/1162381/how-do-i-escape-a-reserved-word-in-oracle):

> [...] Oracle appears to use double quotes (", eg "table") and apparently requires the correct case—whereas, for anyone interested, MySQL defaults to using backticks (`) except when set to use double quotes for compatibility.

### 2. MySQL

In MySQL you quote reserved keywords via the backtick character:

```sql
CREATE TABLE `interval` (begin INT, end INT);
```

Source: [MySQL 8.0 Reference Manual  /  Language Structure  /  Keywords and Reserved Words](https://dev.mysql.com/doc/refman/8.0/en/keywords.html)

### 3. Postgres

Here, it seems to be a simple double quote `"` character:

> There is a second kind of identifier: the delimited identifier or quoted identifier. It is formed by enclosing an arbitrary sequence of characters in double-quotes ("). A delimited identifier is always an identifier, never a key word. So "select" could be used to refer to a column or table named “select”, whereas an unquoted select would be taken as a key word and would therefore provoke a parse error when used where a table or column name is expected. The example can be written with quoted identifiers like this:

Source: [Documentation → PostgreSQL 13 → 4.1. Lexical Structure → 4.1.1. Identifiers and Key Words](https://www.postgresql.org/docs/current/sql-syntax-lexical.html)

### 4. SQL Server

It seems that square brackets `[FIELD]` are used to escape a reserved keyword.

I was not able to find THE one original resource to proof this, but those links might be useful:

[StackOverflow: How to deal with SQL column names that look like SQL keywords?](https://stackoverflow.com/questions/285775/how-to-deal-with-sql-column-names-that-look-like-sql-keywords)

[What is the use of the square brackets [] in sql statements?](https://stackoverflow.com/questions/52898/what-is-the-use-of-the-square-brackets-in-sql-statements)

[Delimited Identifiers (Database Engine)](https://docs.microsoft.com/en-us/previous-versions/sql/sql-server-2008-r2/ms176027(v=sql.105)?redirectedfrom=MSDN)

### 5. Db2

Seems to be the double quote character as well `"`:

> However, a keyword can be used as an identifier in a context where it is a reserved word, by specifying it as a delimited identifier. For example:
> 
> - ALL cannot be a column name in a SELECT statement, unless it is delimited. However, if the quotation mark (") is the escape character that begins and ends delimited identifiers, “ALL” can be used as a column name in a SELECT statement.
> [...]

Source: [Db2 for z/OS > Reserved words](https://www.ibm.com/docs/en/db2-for-zos/11?topic=words-reserved)

### Sidenote

I have only deep experience with MySQL. All other database types are based on documentation reading.

Does this help @gvenzl ?

andygrunwald committed:
Add support for multiple database types

andygrunwald commented:
Hey @gvenzl,

based on my research, I added quote support of column names for all five database types.
Would be nice if you can provide a review.

Let me know what you think

andygrunwald commented:
Hey @gvenzl,
any update on this? Anything I can do to speed it up?

gvenzl commented:
Hey @andygrunwald, my sincere apologies!
Life has caught up with me and I dropped the ball on this until now.

Thanks a lot for your analysis and the PR!

I found that SQL Server seems to support both, the square brackets (`[]`) and the double-quote identifiers (`""`): https://docs.microsoft.com/en-us/sql/t-sql/statements/set-quoted-identifier-transact-sql?view=sql-server-ver15
The latter is thanks to the parameter `QUOTED_IDENTIFIER` which the documentation says is turned `ON` by default but actually, on my test environment, it appears to be `OFF` by default.

The double-quoted identifiers are also what the SQL standard specifies.
The only outlier appears to be MySQL.

However, as stated before, there is another implication with quoted identifiers, which is that they also treat the identifier case-sensitive.
Unfortunately, to make matters a bit more complex, the way how they treat the case is also not consistent

For example, Db2 treats everything in uppercase by default:

```sql
db2 => connect to test;

   Database Connection Information

 Database server        = DB2/LINUXX8664 11.5.4.0
 SQL authorization ID   = DB2INST1
 Local database alias   = TEST

db2 => create table test (rank int);
DB20000I  The SQL command completed successfully.
db2 =>
db2 => create table test1 ("rank" int);
DB20000I  The SQL command completed successfully.
db2 =>
db2 => create table test2 ("RANK" int);
DB20000I  The SQL command completed successfully.
db2 =>
db2 => select rank from test;

RANK
-----------

  0 record(s) selected.

db2 => select rank from test1;
SQL0104N  An unexpected token "" was found following "SYSIBM.RANK".  Expected
tokens may include:  "OVER".  SQLSTATE=42601
db2 =>
db2 => select rank from test2;

RANK
-----------

  0 record(s) selected.

db2 => describe table test;

                                Data type                     Column
Column name                     schema    Data type name      Length     Scale Nulls
------------------------------- --------- ------------------- ---------- ----- ------
RANK                            SYSIBM    INTEGER                      4     0 Yes

  1 record(s) selected.

db2 => describe table test1;

                                Data type                     Column
Column name                     schema    Data type name      Length     Scale Nulls
------------------------------- --------- ------------------- ---------- ----- ------
rank                            SYSIBM    INTEGER                      4     0 Yes

  1 record(s) selected.

db2 => describe table test2;

                                Data type                     Column
Column name                     schema    Data type name      Length     Scale Nulls
------------------------------- --------- ------------------- ---------- ----- ------
RANK                            SYSIBM    INTEGER                      4     0 Yes

  1 record(s) selected.

db2 =>
```

You can see that table `test1` actually has a lowercase specified column named `rank`. Given that the database treats every identifier as uppercase by default, the `select rank from test1;` really becomes a `select RANK from TEST1;` which in that case does not exist.

The same is true for Oracle DB:

```sql
Connected to:
Oracle Database 18c Express Edition Release 18.0.0.0.0 - Production
Version 18.4.0.0.0

SQL> create table test (rank int);

Table created.

SQL> create table test1 ("rank" int);

Table created.

SQL> create table test2 ("RANK" int);

Table created.

SQL> select rank from test;

no rows selected

SQL> select rank from test1;
select rank from test1
       *
ERROR at line 1:
ORA-00904: "RANK": invalid identifier


SQL> select rank from test2;

no rows selected

SQL> describe test;
 Name					   Null?    Type
 ----------------------------------------- -------- ----------------------------
 RANK						    NUMBER(38)

SQL> describe test1;
 Name					   Null?    Type
 ----------------------------------------- -------- ----------------------------
 rank						    NUMBER(38)

SQL> describe test2;
 Name					   Null?    Type
 ----------------------------------------- -------- ----------------------------
 RANK						    NUMBER(38)

SQL>
```

However, Postgres on the other hand treats everything lowercase by default, unlike Oracle or Db2:

```sql
psql (13.4 (Debian 13.4-1.pgdg100+1))
Type "help" for help.

test=> create table test (rank int);
CREATE TABLE
test=> create table test1 ("rank" int);
CREATE TABLE
test=> create table test2 ("RANK" int);
CREATE TABLE
test=> select rank from test;
 rank
------
(0 rows)

test=> select rank from test1;
 rank
------
(0 rows)

test=> select rank from test2;
ERROR:  column "rank" does not exist
LINE 1: select rank from test2;
               ^
test=> \d test;
                Table "public.test"
 Column |  Type   | Collation | Nullable | Default
--------+---------+-----------+----------+---------
 rank   | integer |           |          |

test=> \d test1;
               Table "public.test1"
 Column |  Type   | Collation | Nullable | Default
--------+---------+-----------+----------+---------
 rank   | integer |           |          |

test=> \d test2;
               Table "public.test2"
 Column |  Type   | Collation | Nullable | Default
--------+---------+-----------+----------+---------
 RANK   | integer |           |          |

test=>
```

Yet MySQL does not care about the case of the quoted identifiers:

```sql
Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql> create table test1 (`rank` int);
Query OK, 0 rows affected (0.05 sec)

mysql> create table test2 (`RANK` int);
Query OK, 0 rows affected (0.07 sec)

mysql> select `rank` from test1;
Empty set (0.00 sec)

mysql> select `rank` from test2;
Empty set (0.00 sec)

mysql> describe test1;
+-------+------+------+-----+---------+-------+
| Field | Type | Null | Key | Default | Extra |
+-------+------+------+-----+---------+-------+
| rank  | int  | YES  |     | NULL    |       |
+-------+------+------+-----+---------+-------+
1 row in set (0.01 sec)

mysql> describe test2;
+-------+------+------+-----+---------+-------+
| Field | Type | Null | Key | Default | Extra |
+-------+------+------+-----+---------+-------+
| RANK  | int  | YES  |     | NULL    |       |
+-------+------+------+-----+---------+-------+
1 row in set (0.01 sec)
```

This is also noted in the [MySQL documentation](https://dev.mysql.com/doc/refman/8.0/en/identifier-case-sensitivity.html):

> Partition, subpartition, column, index, stored routine, event, and resource group names are not case-sensitive on any platform, nor are column aliases.

SQL Server, btw, does also not seem to care much about the case-sensitivity of the quoted identifier:

```sql
1> create table test ([rank] int);
2> go
1> create table test1 ([rank] int);
2> go
1> create table test2 ([RANK] int);
2> go
1> select rank from test;
2> go
rank
-----------

(0 rows affected)
1> select rank from test1;
2> go
rank
-----------

(0 rows affected)
1> select rank from test2;
2> go
rank
-----------

(0 rows affected)
1> sp_help test;
2> go
Name                                                                                                                             Owner                                                                                                                            Type                            Created_datetime
-------------------------------------------------------------------------------------------------------------------------------- -------------------------------------------------------------------------------------------------------------------------------- ------------------------------- -----------------------
test                                                                                                                             dbo                                                                                                                              user table                      2021-08-17 02:51:56.267


Column_name                                                                                                                      Type                                                                                                                             Computed                            Length      Prec  Scale Nullable                            TrimTrailingBlanks                  FixedLenNullInSource                Collation
-------------------------------------------------------------------------------------------------------------------------------- -------------------------------------------------------------------------------------------------------------------------------- ----------------------------------- ----------- ----- ----- ----------------------------------- ----------------------------------- ----------------------------------- --------------------------------------------------------------------------------------------------------------------------------
rank                                                                                                                             int                                                                                                                              no                                            4 10    0     yes                                 (n/a)                               (n/a)                               NULL

Identity                                                                                                                         Seed                                     Increment                                Not For Replication
-------------------------------------------------------------------------------------------------------------------------------- ---------------------------------------- ---------------------------------------- -------------------
No identity column defined.                                                                                                                                          NULL                                     NULL                NULL

RowGuidCol
--------------------------------------------------------------------------------------------------------------------------------
No rowguidcol column defined.

Data_located_on_filegroup
--------------------------------------------------------------------------------------------------------------------------------
PRIMARY

The object 'test' does not have any indexes, or you do not have permissions.

No constraints are defined on object 'test', or you do not have permissions.

No foreign keys reference table 'test', or you do not have permissions on referencing tables.
No views with schema binding reference table 'test'.
1> sp_help test1;
2> go
Name                                                                                                                             Owner                                                                                                                            Type                            Created_datetime
-------------------------------------------------------------------------------------------------------------------------------- -------------------------------------------------------------------------------------------------------------------------------- ------------------------------- -----------------------
test1                                                                                                                            dbo                                                                                                                              user table                      2021-08-17 02:52:01.017


Column_name                                                                                                                      Type                                                                                                                             Computed                            Length      Prec  Scale Nullable                            TrimTrailingBlanks                  FixedLenNullInSource                Collation
-------------------------------------------------------------------------------------------------------------------------------- -------------------------------------------------------------------------------------------------------------------------------- ----------------------------------- ----------- ----- ----- ----------------------------------- ----------------------------------- ----------------------------------- --------------------------------------------------------------------------------------------------------------------------------
rank                                                                                                                             int                                                                                                                              no                                            4 10    0     yes                                 (n/a)                               (n/a)                               NULL

Identity                                                                                                                         Seed                                     Increment                                Not For Replication
-------------------------------------------------------------------------------------------------------------------------------- ---------------------------------------- ---------------------------------------- -------------------
No identity column defined.                                                                                                                                          NULL                                     NULL                NULL

RowGuidCol
--------------------------------------------------------------------------------------------------------------------------------
No rowguidcol column defined.

Data_located_on_filegroup
--------------------------------------------------------------------------------------------------------------------------------
PRIMARY

The object 'test1' does not have any indexes, or you do not have permissions.

No constraints are defined on object 'test1', or you do not have permissions.

No foreign keys reference table 'test1', or you do not have permissions on referencing tables.
No views with schema binding reference table 'test1'.
1> sp_help test2;
2> go
Name                                                                                                                             Owner                                                                                                                            Type                            Created_datetime
-------------------------------------------------------------------------------------------------------------------------------- -------------------------------------------------------------------------------------------------------------------------------- ------------------------------- -----------------------
test2                                                                                                                            dbo                                                                                                                              user table                      2021-08-17 02:52:06.880


Column_name                                                                                                                      Type                                                                                                                             Computed                            Length      Prec  Scale Nullable                            TrimTrailingBlanks                  FixedLenNullInSource                Collation
-------------------------------------------------------------------------------------------------------------------------------- -------------------------------------------------------------------------------------------------------------------------------- ----------------------------------- ----------- ----- ----- ----------------------------------- ----------------------------------- ----------------------------------- --------------------------------------------------------------------------------------------------------------------------------
RANK                                                                                                                             int                                                                                                                              no                                            4 10    0     yes                                 (n/a)                               (n/a)                               NULL

Identity                                                                                                                         Seed                                     Increment                                Not For Replication
-------------------------------------------------------------------------------------------------------------------------------- ---------------------------------------- ---------------------------------------- -------------------
No identity column defined.                                                                                                                                          NULL                                     NULL                NULL

RowGuidCol
--------------------------------------------------------------------------------------------------------------------------------
No rowguidcol column defined.

Data_located_on_filegroup
--------------------------------------------------------------------------------------------------------------------------------
PRIMARY

The object 'test2' does not have any indexes, or you do not have permissions.

No constraints are defined on object 'test2', or you do not have permissions.

No foreign keys reference table 'test2', or you do not have permissions on referencing tables.
No views with schema binding reference table 'test2'.
1>
```

```sql
1> set quoted_identifier on;
2> go
1> create table test (rank int);
2> go
1> create table test1 ("rank" int);
2> go
1> create table test2 ("RANK" int);
2> go
1> select rank from test;
2> go
rank
-----------

(0 rows affected)
1> select rank from test1;
2> go
rank
-----------

(0 rows affected)
1> select rank from test2;
2> go
rank
-----------

(0 rows affected)
1>
```

gvenzl commented:
I think that the conclusion of all of this is that if we introduce quoted identifier support into `csv2db`, it will have to be via an additional parameter, not just blank over everything.

There is still an outstanding question of what the quote characters should be for SQL Server, the legacy `[]` or the standard, yet apparently not turned on by default `""`.

Last but not least, identifiers also apply to table names.
If one would want to quote identifiers, it is to be assumed that it should apply for both, the column names, as well as the table name itself.

gvenzl commented on README.md:
````diff
@@ -192,7 +195,7 @@ Closing database connection.
 
 ```sql
-$ ./csv2db generate -f test/resources/201811-citibike-tripdata.csv
+$ ./csv2db generate -f test/resources/201811-citibike-tripdata.csv -o mysql
````
I don't think this should change in the ReadMe. For the majority of use cases, running without the database type should be sufficient to generate the table DDL

andygrunwald replied:
Thats fair. Sorry for changing this. Fixed in https://github.com/csv2db/csv2db/pull/51/commits/5a890c37c8575a6b6d95bc3f418fe8ef682e672f.

gvenzl commented on README.md:
````diff
@@ -216,7 +219,7 @@ CREATE TABLE <TABLE NAME>
 
 ```sql
-$ ./csv2db generate -f test/resources/201811-citibike-tripdata.csv -t STAGING
+$ ./csv2db generate -f test/resources/201811-citibike-tripdata.csv -t STAGING -o mysql
````
I don't think this should change in the ReadMe. For the majority of use cases, running without the database type should be sufficient to generate the table DDL

andygrunwald replied:
Thats fair. Sorry for changing this. Fixed in https://github.com/csv2db/csv2db/pull/51/commits/5a890c37c8575a6b6d95bc3f418fe8ef682e672f.

gvenzl commented on README.md:
````diff
@@ -240,7 +243,7 @@ CREATE TABLE STAGING
 
 ```sql
-$ ./csv2db generate -f test/resources/201811-citibike-tripdata.csv -t STAGING -c CLOB
+$ ./csv2db generate -f test/resources/201811-citibike-tripdata.csv -t STAGING -c CLOB -o mysql
````
I don't think this should change in the ReadMe. For the majority of use cases, running without the database type should be sufficient to generate the table DDL

andygrunwald replied:
Thats fair. Sorry for changing this. Fixed in https://github.com/csv2db/csv2db/pull/51/commits/5a890c37c8575a6b6d95bc3f418fe8ef682e672f.

andygrunwald committed:
Remove `-o mysql` from README

andygrunwald commented:
@gvenzl Thanks a lot for this big analysis. Let's break down the challenges followed by suggestions from me:

### Quoted identifier via an additional parameter

That is a very good idea, I agree. This will not change the current behavior, but also add the possibility for all.
Let me work on this and modify this PR.

- [ ] Add an additional parameter to quote fields and tables

### Quote character (`"` vs. `[]`)

What do you think about using `"` as default and adding yet another parameter to switch the quote character to `[]`.
By this, the user can switch, because they know their setup best.

### Quoting table names

Good point. Thanks for catching this.
Here, I agree as well. Table names should also be quotes once enabled.
Let me work on this and modify this PR.

- [ ] When quoting is enabled, also quote table names

### Lower vs. Uppercase of database systems

Maybe it is useful to document this knowledge because it seems to be applied once identifiers are quoted.
A table in the README can help here.

### Your feedback

What do you think?
If these changes are applied, do you see any chance of adding this into the main branch?

gvenzl commented:
Thanks @andygrunwald for breaking it down!

I agree with all the points you have made.
For the `[]` vs `""` I still want to research a bit more what SQL Server actually recommends.
Adding a separate parameter for just that strikes me as a bit too excessive.
It could be that, for example, it turns out that `[]` is already deprecated/discouraged.
And if not, `csv2db` could perhaps be smart enough to check for the parameter at runtime and make the decision itself (if that parameter value is accessible to non-DBA database users that a user will most likely load data with).

Yeah, I think it we have tracked it all down, we can merge it into main.

gvenzl commented:
We should also agree on the parameter name.
I think `--quote-identifiers` makes it quite explicit.
However, I just saw th`-q` (`--quote`) is already occupied for the string quotation of the actual data.
Perhaps we don't need a shorthand parameter for this?

andygrunwald commented:
Personally, I am a fan of being explicit and try to use long parameter names.
I am fine with `--quote-identifiers`, but also `--use-quote-identifiers` would work as well. Or `--keyword-quoting` or something like this.
I am also OK with dropping a shorthand. 

Let me try to reserve some time in the next days to get something done. I will ping you for early reviews. Ok?

gvenzl commented:
Great! `quote` in the case of `--quote-identifiers` is a verb, i.e. "Please csv2db, quote these identifiers for me".
For the `use`, technically it should be `--use-quoted-identifiers`, as they are generally referred to as "quoted identifiers". "Keyword" is technically not correct, as a keyword is different from an identifier.
Consider, for example, `SELECT col1 FROM table1;`: both `col1` and `table1` are identifiers, while `SELECT` and `FROM` are keywords.

Given your comments, I would say let's please go with `--quote-identifiers`.

andygrunwald commented:
Fair enough and good point. Thanks. Lets go for it.

gvenzl commented:
Awesome, thanks!

andygrunwald committed:
Add TODO to decide which SQL-Server quote character we use

andygrunwald committed:
Rename function "quote_field" to "quote"

Main reason:
We aim not only to quote fields, but also table names.

andygrunwald committed:
Add TODO to follow up on the quoting character for MSSQL/SQL-Server

andygrunwald committed:
Quoting table names

andygrunwald committed:
Introduce a new flag: --quote-identifiers

This flag will quote identifiers like database table names
and database columns. Default is false.

andygrunwald committed:
Add docs about --quote-identifiers and case sensitivity of database systems

andygrunwald commented on src/python/csv2db.py:
```diff
@@ -53,6 +53,7 @@ def run(cmd):
 
     # Set table name
     cfg.table_name = args.table
+    # TODO Should we quote the table name in the debug output as well, once this is configured?
     f.debug("Table name: {0}".format(cfg.table_name))
```
@gvenzl What do you think about this?

andygrunwald commented:
@gvenzl I spent some time applying the changes we have discussed.
Read through it and let me know what you think:

### Quoted identifier via an additional parameter

I introduced a new flag `--quote-identifiers` (see https://github.com/csv2db/csv2db/pull/51/commits/0deba4aa466a1df9a41737604edfc2133ee87377) and updated the docs (see https://github.com/csv2db/csv2db/pull/51/commits/0707922cdc6921020092b5472d9678c80c8d09c8).

### Quote character for SQL-Server (`"` vs. `[]`)

A bit more research needs to be done on this topic.
I am not that familiar with SQL-Server.

For now, I added a TODO to decide which SQL-Server quote character we use.
See https://github.com/csv2db/csv2db/pull/51/commits/301cb0c67adb3babb4bc107af2c461eec1bd4534

The question remains on how we progress here on this.
How can I help? Do you plan to do it?

One idea can be to document the limitation, create an issue to fix this.
This would have the benefit of not blowing up this PR further and not blocking it any longer.
What do you think? 

### Quoting table names

Done in https://github.com/csv2db/csv2db/pull/51/commits/63dc1da9b2734c672d921ee9b6a3e9eedf06aefe

### Lower vs. Uppercase of database systems

As discussed, I documented the behavior in https://github.com/csv2db/csv2db/pull/51/commits/0707922cdc6921020092b5472d9678c80c8d09c8

### Open topics

#### Unit/Integration tests

We don't have any automated tests yet.
Personally, I would prefer to wait for an easier test solution like described in https://github.com/csv2db/csv2db/issues/52

### Feedback

That's it from my side.
Anything you have in mind what should be tackled?

andygrunwald committed:
Fix whitespace at the end of the lines

andygrunwald committed:
Fix license whitespace

gvenzl commented:
Superseded by #61
